xcodeproj/internal/linker_input_files.bzl

```diff
@@ -21,84 +22,189 @@ def _collect(*, cc_info, objc, is_xcode_target):
+            `linker_input_files.collect_for_top_level`. These inputs will be
+            excluded from the return list.
+
+    Returns:
+        A value similar to the one returned from
+        `linker_input_files.collect_for_non_top_level`.
+    """
+    return _merge(deps = deps, avoid_linker_inputs = avoid_linker_inputs)
+
+def _merge(*, deps, avoid_linker_inputs = None):
     """Merges linker input files from the deps of a target.
 
-    This should only be used by top level targets (which don't have access to
-    their own linker input files), or targets that are being skipped.
+    This should only be used by targets that are being skipped.
 
     Args:
         deps: `ctx.attr.deps` of the target.
+        avoid_linker_inputs: A value returned from
+            `linker_input_files.collect_for_top_level`. These inputs will be
+            excluded from the return list.
 
     Returns:
-        A value similar to the one returned from `linker_input_files.collect`.
+        A value similar to the one returned from
+        `linker_input_files.collect_for_top_level`.
     """
-    transitive_linker_inputs = [
-        (dep[XcodeProjInfo].target, dep[XcodeProjInfo].linker_inputs)
-        for dep in deps
-        if dep[XcodeProjInfo].linker_inputs
-    ]
-
     cc_linker_inputs = depset(
         order = "topological",
         transitive = [
-            linker_inputs._cc_linker_inputs
-            for _, linker_inputs in transitive_linker_inputs
+            dep[XcodeProjInfo].linker_inputs._cc_linker_inputs
+            for dep in deps
+        ],
+    )
+
+    dynamic_framework_files = depset(
+        order = "topological",
+        transitive = [
+            dep[XcodeProjInfo].linker_inputs._dynamic_framework_files
+            for dep in deps
         ],
     )
 
     static_framework_files = depset(
         order = "topological",
         transitive = [
-            linker_inputs._static_framework_files
-            for _, linker_inputs in transitive_linker_inputs
+            dep[XcodeProjInfo].linker_inputs._static_framework_files
+            for dep in deps
         ],
     )
 
     xcode_library_targets = [
-        target
-        for target, linker_inputs in transitive_linker_inputs
-        if linker_inputs._is_xcode_library_target
+        dep[XcodeProjInfo].target
+        for dep in deps
+        if dep[XcodeProjInfo].linker_inputs._is_xcode_library_target
     ]
 
     return struct(
+        _avoid_linker_inputs = avoid_linker_inputs,
         _cc_linker_inputs = cc_linker_inputs,
+        _dynamic_framework_files = dynamic_framework_files,
         _static_framework_files = static_framework_files,
         _is_xcode_library_target = False,
         xcode_library_targets = xcode_library_targets,
     )
 
-def _get_files_to_link(linker_inputs):
-    """Returns the files to link for a target.
+def _get_static_libraries(linker_inputs):
+    """Returns the static libraries needed to link the target.
```

Comment:
I want to remove this in the future. It's duplicative of work done in `_to_dto`, and it's only used by the `required_merges` calculations.